### PR TITLE
[Ready] Adds a new RnD method, the B.E.P.I.S.

### PR DESCRIPTION
**Pull request by ArcaneMusic**
This is ArcaneMusic's feature number **2** designed to force you to squirrel out your hard earned cash in a round in a constant effort to give money real value within a round. All values here are tentative, up for discussion, and right now are meant to give a vastly higher value payout at higher levels, subject to a degree of random chance.

## About The Pull Request
**The BEPIS, what is it, and how does it work?**
I'm glad you asked! The BEPIS, or **B**usiness **E**xploration **P**rotocol **I**ncubation **S**ink, is a machine meant to sit nestled right next to the E.X.P.E.R.I-MENTOR in science. Think of the two machines as brother and sister, actually. If the E.X.P.E.R.I-MENTOR uses brute force methods to discern the nature of objects and phenomena, then the B.E.P.I.S. takes a gentler approach, by committing tons of resources to the problem so that someone else will either figure it out, or give you an excuse to fire them. 
image
_The B.E.P.I.S. and 6 of it's possible, beneficial rewards._

Specifically, the B.E.P.I.S. runs off of 2 main resources, Money, in all its shapes and sizes, and Electricity, which during a long shift are the two things the station should have in DROVES. How the machinery works is you begin by inserting money into the machine, for example, 7000 credits. Then, based on Gaussian distributions, your investment will either increase in value due to it having research merit, or fall due to being an exceptionally terrible idea but hey, you're the one funding it so whatever. If you pass either the major or minor threshold, the B.E.P.I.S. will produce either individual products, or technology disks allowing you to reveal new techs within the tech tree.

**THE TECHS TO BE ADDED TO THE TREE ARE NOT INTENDED TO BLOCK OFF OLD, EXISTING TECHS.**
It's JUST for interesting, but otherwise not irreplaceable items that serve as a fun mid-end game tech tree additions.
If you obtain a Major Investment threshold, right now the only prize is a tech disk containing the Illumination Applications,  which is unlocked via the deconstruction analyzer, allowing access to a smaller version of the RLD and a new, upgraded version of the fire helmet.
The Minor Investment threshold rewards include a new way of painting on decals onto floor tiles, the Decal Painter (The blue airlock painter), and the Polycircuit Aggregate, which acts as a functional adaptive psudocircuit which if interrupted will slice you up because it's a nightmare of circuits thrown together into a mess.

In the off chance you don't meet either threshold, you'll get a funny reason for why your venture capitol project didn't pan out. It pays to invest deeply, if your goal is to get a reward.
In the even more unlikely chance that you not only don't meet either threshold, but you somehow LOSE money in the process, The machine will malfunction, and cause a power surge, burning through a lot of power in the APC. Also something about bucks?
image
Breakdown of the UI and the variables on the bottom of the screen.
_Deposited Credits:_ Money currently being risked by testing.
_Investment Variability:_ How wide of a margin your money can fluctuate both positively and negatively. Upgrading makes your investment more consistent. 
_Innovation Bonus:_ Provides a static bonus to the value of any test.
_Risk offset:_ Decreases the cost of both the Minor and Major reward linearly.
_Average Technology Cost:_ The mean, raw value needed to obtain a new tech disk. NOTE: Because this uses Gaussian distribution, this is the point where your odds are 50% even, NOT guaranteed.

Quick aside, I also wound up running into an issue that **Skoglol** helped me out with, turns out that the copy_research_to proc was written in a way so that it only really added the tech and the designs from a disk if you uploaded the tech twice, so that's been re-ordered to future-proof that for the next time hidden nodes get added.
## Why It's Good For The Game
This has been an idea I've been dreaming up for awhile now, with three expressed goals:
  A.) Give the Science Budget something to spend they money towards.
  B.) Create a Money Sink in the Science department, which fits with the theme of NT being absolutely nuts about their bottom line and making a profit.
  C.) Give an excuse for there to be more diverse methods of gaining hidden tech nodes within a single round.
  At least until someone makes a way for heads to buy crates straight from their department budget within their own department, I'll consider this a win in terms of A and B. This also at least lays the groundwork for C, as other people will be free to add more minor and major rewards and techs.

## To-Do List:

- [x] Add the basic functionality.
- [x] Add at least one major, and one minor reward for using the B.E.P.I.S.
- [x] Set up an easy framework for being able to add new methods of unlocking new, blocked techs. (Close enough for now.)
- [x] Set up machine upgrades to improve odds/Show odds of success (Ugh, MATH).
- [x] Maybe add another tech/ a few extra minor rewards.
- [x] Add the B.E.P.I.S. to all or most map's RnD departments.
- [x] Bucks?

## Changelog
:cl:
add: The RnD department has been gifted a state-of-the-art new research tool, the B.E.P.I.S. by Nanotrasen's Venture Capital Labs, after repeatedly wasting all of their funds. Hopefully you'll find a... wiser use for it.
add: A new Tech Node can be unlocked through proper investment with the B.E.P.I.S.
add: In an unrelated piece of news, a deer was recently found to be the culprit of Nanotrasen Venture Capital Labs's onboard nuclear failsafe device detonating. Yes, we're sure it wasn't the Syndicate this time. Yes, we wish we were joking too.
fix: Fixes a small bug with tech disks containing hidden nodes not researching.
/:cl:

### CODE REVIEW

**Comment by HomocidlePancakes** [comment]
BEPIS :DDDDDDDDD

**Comment by TheVekter** [comment]
This is a neat idea. I feel like we'll need to come up with a fair few cool things for this to make or no one's gonna build it just for a better RLD and fire helmet. I'm honestly perfectly okay with locking some of the higher research stuff behind this, but I can see why that would be an unpopular opinion.

**Comment by cacogen** [comment]
1. bad name
1. wip pull no allow
2. looks like a risk of rain sprite without the outlines
3. why does it have to be a huge tank if nobody's going inside it, all it needs is a screen and a holochip scanner
4. like the experimentor, no one will use it, especially if at the moment it just gives you a new node and fire helmet with promises of other stuff to be added in future
5. why not just make research points and money interchangeable? aren't they the same thing? bomb recipes grant research points in direct proportion to their monetary value

i like the satire of investing in awful start-ups though, which gives it a foot in the real world

**Comment by SkeletalElite** [comment]
> 2. wip pull no allow

They have to be done enough to be at least test mergeable on the live server, which it seems this one is

**Comment by ArcaneMusic** [comment]
> 1. bad name
> 2. wip pull no allow
> 3. looks like a risk of rain sprite without the outlines
> 4. why does it have to be a huge tank if nobody's going inside it, all it needs is a screen, holochip reader and lathe
> 5. like the experimentor, no one will use it, especially if at the moment it just gives you a new node and fire helmet with promises of other stuff to be added in future
> 6. why not just make research points and money interchangeable? aren't they the same thing? bomb recipes grant research points in direct proportion to their monetary value
> 
> i like the satire of investing in awful start-ups though, which gives it a foot in the real world

1.) The name can be changed. I'm not married to it, and I've asked for feedback on it multiple times.
2.) The feature functionally works, it's not a draft. I just know it's not feature complete but it's in a state where I can take feedback, which is why we're here.
4.) For effect.
5.) I'm working on at least one more additional tech for it right now, the only issue is that I am a single person who can only do the work of a single person.
6.) As it stands, the balance of money and RnD points is incredibly slanted because of RnD bombs. The only way to pay for investments at the moment is out of pocket by members of science, due to not being able to touch the science budget. Being able to just straight up buy RnD Points without an element of risk is not my goal, and I'm working on additional content to fill out the Minor Threshold reward table at the moment, so that pumping money into RnD doesn't allow you to Max out RnD in 10 minutes.

**Comment by Krysonism** [comment]
Great PR, i have some ideas for minor rewards or if they are too stupid to be considered a reward they can be generated on failure.

File: code/modules/research/bepis.dm [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+			else if(success_type == 3)
+				say("ERROR: CRITICAL MACHIME MALFUNCTI- ON. CURRENCY IS NOT CRASH. CANNOT COMPUTE COMMAND: 'make bucks'") //not a typo, for once.
+				banked_cash = 0
+				new /mob/living/simple_animal/deer(src.loc, 1)
+				use_power(50000) //To prevent gambling at low cost and also spawming for infinite deer.
+				flick("chamber_flash",src)
+				chamber_status = 1
+				update_icon_state()
+			success_type = 0
+	if("amount")
```
Comment (spookydonut) [inline code/modules/research/bepis.dm]: I dont think this is indented enough

**Review by spookydonut — CHANGES_REQUESTED** [review]

File: code/game/machinery/doors/airlock_electronics.dm [hunk 2]
```diff
@@ -65,3 +65,51 @@
 			var/unres_direction = text2num(params["unres_direction"])
 			unres_sides ^= unres_direction //XOR, toggles only the bit that was clicked
 			. = TRUE
+
+/obj/item/stack/circuit_stack
+	name = "polycircuit aggregate"
+	desc = "A dense, overdesigned cluster of electronics which attempted to function as a multipurpose circuit electronic. Circuits can be removed from it... ff you don't bleed out in the process."
```
Comment (BadSS13Player) [inline code/game/machinery/doors/airlock_electronics.dm]: >ff

spelling error

**Comment by Iamgoofball** [comment]
Make it generate a sales pitch that is then blasted out with say() so we can put it on the radio.

**Comment by Shadowflame909** [comment]
The heck is an RLD?

**Comment by ArcaneMusic** [comment]
> The heck is an RLD?

A full RLD spawns on runtimestation, but basically it has 3 modes: Creating light fixtures, removing light fixtures, or being able to fire glowsticks. Both the glowsticks and the light fixtures can have their color set as well. All of these take metal/glass/materials, just like an RCD.


> Make it generate a sales pitch that is then blasted out with say() so we can put it on the radio.


Solid maybe, if I can figure out how to do it a more elegant way than vending machines do it.

**Review by stylemistake — DISMISSED** [review]

**Comment by ArcaneMusic** [comment]
At oranges suggestion I ditched the RnD points reward, considering that I realized that could easily get gamed to give you a metric ton of points potentially alongside medical surgeries. (Research notes when combined with another note from a difference "source" id multiplies the value of the reward)

File: code/game/machinery/doors/airlock_electronics.dm [hunk 3]
```diff
@@ -65,3 +65,53 @@
...
+
+/obj/item/stack/circuit_stack/attack_self(mob/user)// Prevents the menu, and tells you how to use it.
+	to_chat(user, "<span class='warning'>You can't use the [src] by itself, you'll have to try and remove one of these circuits by hand... carefully.</span>")
+
+/obj/item/stack/circuit_stack/attack_hand(mob/user)
+	var/mob/living/carbon/human/H = user
+	if(user.get_inactive_held_item() == src)
+		if(zero_amount())
+			return
+		chosen_circuit = input("What type of circuit would you like to remove?", "Choose a Circuit Type", chosen_circuit) in list("airlock","firelock","fire alarm","air alarm","APC")
```
Comment (AnturK) [inline code/game/machinery/doors/airlock_electronics.dm]: Add amount check after the input.

File: code/game/objects/items/airlock_painter.dm [hunk 4]
```diff
@@ -125,3 +125,112 @@
...
+	custom_materials = list(/datum/material/iron=2000, /datum/material/glass=500)
+	var/stored_dir = 2
+	var/stored_color = ""
+	var/stored_decal = "warningline"
+	var/stored_decal_total = "warningline"
+
+/obj/item/airlock_painter/decal/afterattack(atom/target, mob/user, proximity)
+	. = ..()
+	var/turf/open/floor/F = target
+	if(!(proximity))
```
Comment (AnturK) [inline code/game/objects/items/airlock_painter.dm]: drop the redundant ()'s please

File: code/game/objects/items/airlock_painter.dm [hunk 5]
```diff
@@ -125,3 +125,112 @@
...
+		to_chat(user, "<span class='notice'>You need to get closer!</span>")
+		return
+	if(use_paint(user) && isturf(F))
+		F.AddComponent(/datum/component/decal, 'icons/turf/decals.dmi', stored_decal_total, stored_dir, CLEAN_STRONG, color, null, null, alpha) //Make sure to delete in final version.
```
Comment (AnturK) [inline code/game/objects/items/airlock_painter.dm]: Delete what

File: code/game/objects/items/airlock_painter.dm [hunk 6]
```diff
@@ -125,3 +125,112 @@
...
+		if("yellow")
+			stored_color = ""
+			update_decal_path()
+		if("red")
+			stored_color = "_red"
+			update_decal_path()
+		if("white")
+			stored_color = "_white"
+			update_decal_path()
+			. = TRUE
```
Comment (AnturK) [inline code/game/objects/items/airlock_painter.dm]: Why only this one causes update, i assume you want it on all. Also you could reduce copypaste here by doing update_decal path at the end of the switch.

File: code/modules/research/bepis.dm [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+	var/minor_threshold = 3000
+	var/std = 1000 //That's Standard Deviation, what did you think it was?
+	//Stock part variables
+	var/power_saver = 1
+	var/inaccuracy_percentage = 1.5
+	var/positive_cash_offset = 0
+	var/negative_cash_offset = 0
+
+/obj/machinery/rnd/bepis/attackby(obj/item/O, mob/user, params)
+	if(powered == FALSE)
```
Comment (AnturK) [inline code/modules/research/bepis.dm]: Why are you not using common machinery stat.

File: code/modules/research/bepis.dm [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+		to_chat(user, "<span class='notice'>The [src] can't accept money with no power.</span>")
+		return
+	if(istype(O, /obj/item/holochip))
+		var/obj/item/holochip/cash = O
+		var/deposit_value = cash.get_item_credit_value()
```
Comment (AnturK) [inline code/modules/research/bepis.dm]: Copyaste here and below put this in a proc, or collapse into single if

File: code/modules/research/bepis.dm [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+/obj/machinery/rnd/bepis/proc/depositcash()
+	var/deposit_value = 0
+	if(!Card)
+		say("No account detected.")  //No freeloading off of science.
+		return
+	else if(Card.registered_account)
+		account = Card.registered_account
+
+	deposit_value = banking_amount
+	if(deposit_value == null || deposit_value == 0)
```
Comment (AnturK) [inline code/modules/research/bepis.dm]: Why would it ever be null here.

File: code/modules/research/bepis.dm [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+		return
+	else
+		account.adjust_money(-deposit_value) //The money vanishes, not paid to any accounts.
+		banked_cash += deposit_value
+		use_power(1000 * power_saver)
+		say("Cash deposit successful. There is [banked_cash] in the chamber.")
+
+	if(banked_cash >= 1)
+		chamber_status = 2
+		update_icon_state()
```
Comment (AnturK) [inline code/modules/research/bepis.dm]: You didn't call this in attackby set

File: code/modules/research/bepis.dm [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+	density = TRUE
+	use_power = IDLE_POWER_USE
+	idle_power_usage = 1500
+	circuit = /obj/item/circuitboard/machine/bepis
+
+	var/banking_amount = 100
+	var/banked_cash = 0					//stored player cash
+	var/datum/bank_account/account		//payer's account.
+	var/mob/living/carbon/human/H   	//the person using the console in each instance.
+	var/obj/item/card/id/Card   		//the account of the person using the console.
```
Comment (AnturK) [inline code/modules/research/bepis.dm]: These two variables do not need to exist, they're only used in ui_act from what i can see.

File: code/modules/research/bepis.dm [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+
+/obj/machinery/rnd/bepis/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "bepis", name, 500, 400, master_ui, state)
+		ui.open()
+	if(ishuman(user))
+		H = user
+		Card = H.get_idcard(TRUE)
+		update_icon_state()
```
Comment (AnturK) [inline code/modules/research/bepis.dm]: update_icon doesn't seem to depend on either of these variables, you're calling it unconditionally later and these two seem unnecessary,

File: code/modules/research/bepis.dm [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+	RefreshParts()
+	update_icon_state()
+
+/obj/machinery/rnd/bepis/ui_data()
+	var/list/data = list()
+	var/storedcash = banked_cash
```
Comment (AnturK) [inline code/modules/research/bepis.dm]: Unnecessary variable

File: code/modules/research/bepis.dm [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+				dropturf = drop_location()
+			if(powered == FALSE)
+				return
+			calcsuccess()
+			if(banked_cash == 0)
+				say("Please deposit funds to begin testing.")
+				return
+			use_power(10000 * power_saver) //This thing should eat your APC battery if you're not careful.
+			//For assigning major rewards, the tech disk will have a random tech with the bepis_tech variable true, check _techweb.dm
+			if(success_type == 2)
```
Comment (AnturK) [inline code/modules/research/bepis.dm]: This success_type variable seems completely temporary too, return value from calc success instead.

File: code/modules/research/bepis.dm [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+				error_cause = pick("attempted to sell grey products to American dominated market.","attempted to sell gray products to British dominated market.","placed wild assumption that PDAs would go out of style.","simulated product #76 damaged brand reputation mortally.","simulated business model resembled 'pyramid scheme' by 98.7%.","product accidently granted override access to all station doors.")
+				say("Experiment concluded with zero product viability. Cause of error: [error_cause]")
+				banked_cash = 0
+				flick("chamber_flash",src)
+				update_icon_state()
+
+			else if(success_type == 3)
+				say("ERROR: CRITICAL MACHIME MALFUNCTI- ON. CURRENCY IS NOT CRASH. CANNOT COMPUTE COMMAND: 'make bucks'") //not a typo, for once.
+				banked_cash = 0
+				new /mob/living/simple_animal/deer(src.loc, 1)
```
Comment (AnturK) [inline code/modules/research/bepis.dm]: dropturf instead of src.loc ?

File: code/modules/research/bepis.dm [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+		if("toggle_power")
+			if(powered == FALSE)
+				powered = TRUE
+				idle_power_usage = 1500
+				update_icon_state()
+			else
+				powered = FALSE
+				idle_power_usage = 0
+				update_icon_state()
+			. = TRUE
```
Comment (AnturK) [inline code/modules/research/bepis.dm]: Same thing as before, mixed indent on return.

File: code/modules/research/techweb/_techweb.dm [hunk 17]
```diff
@@ -111,17 +125,17 @@
 	modify_point_list(l)
 
 /datum/techweb/proc/copy_research_to(datum/techweb/receiver, unlock_hidden = TRUE)				//Adds any missing research to theirs.
+	if(unlock_hidden)
```
Comment (AnturK) [inline code/modules/research/techweb/_techweb.dm]: Why was this moved here ?

File: code/modules/research/techweb/_techweb.dm [hunk 18]
```diff
@@ -48,6 +50,18 @@
 	id = "SCIENCE"
 	organization = "Nanotrasen"
 
+/datum/techweb/bepis	//Should contain only 1 BEPIS tech selected at random.
+	id = "SCIENCE"
+	organization = "Nanotrasen R&D"
+
+/datum/techweb/bepis/New()
+	. = ..()
+	var/bepis_id = pick(experimental_nodes)	//To add a new tech to the BEPIS, add the ID to this pick list.
```
Comment (AnturK) [inline code/modules/research/techweb/_techweb.dm]: I fail to see the need for these to be separate from hidden nodes. They're functionally identical.

File: code/modules/research/techweb/_techweb.dm [hunk 19]
```diff
@@ -48,6 +50,18 @@
 	id = "SCIENCE"
 	organization = "Nanotrasen"
 
+/datum/techweb/bepis	//Should contain only 1 BEPIS tech selected at random.
+	id = "SCIENCE"
```
Comment (AnturK) [inline code/modules/research/techweb/_techweb.dm]: I'm not sure id's are used for anything but it probably shouldn't be the same as default research web.

**Review by AnturK — CHANGES_REQUESTED** [review]
Quick glance over.

File: code/modules/research/techweb/_techweb.dm (hunk 17 above)
Comment (ArcaneMusic) [reply code/modules/research/techweb/_techweb.dm]: Originally it was at the bottom of copy_research_to, so if you had a tech disk with JUST a hidden tech on it, you had to upload the techs contained within the disk twice before you were given that tech.

**Comment by Timonkeyn** [comment]
Reminder that the only budget on the station is the cargo budget because oranges thought it was a good idea to remove the others

**Comment by Iamgoofball** [comment]
What are you talking about? That's not trye at all

**Comment by ArcaneMusic** [comment]
Alright, that should be all the review changes. As far as my testing goes, everything's in functional working order.

**Comment by ArcaneMusic** [comment]
Alright with that, that should be all of the review changes, and a few bug fixes so that everything looks and acts as smoothly as it should. All in all, as far as I can tell it's working as intended.

File: code/modules/research/bepis.dm (hunk 11 above)
Comment (AnturK) [reply code/modules/research/bepis.dm]: This is still a thing and you added a third one.

Comment (AnturK) [reply code/modules/research/bepis.dm]: In case you're not sure how to remove these:
You have access to the user in both ui_act and ui_data, get the data there instead of storing it on the machine.

**Comment by AnturK** [comment]
Merge conflict and there's still issues left from my previous review.

**Comment by kriskog** [comment]
Merge conflict

File: code/modules/research/bepis.dm [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+			depositcash()
+			update_icon_state()
+		if("begin_experiment")
+			if(powered == FALSE)
+				return
+			if(banked_cash == 0)
+				say("Please deposit funds to begin testing.")
+				return
+			calcsuccess()
+			use_power(10000 * power_saver) //This thing should eat your APC battery if you're not careful.
```
Comment (kriskog) [inline code/modules/research/bepis.dm]: This is in Joule, so 10000 is a lot less than you think it is. The standard APC cell has a charge of 2500000J
Also magic number, make it a define.

File: code/modules/research/bepis.dm [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+		if(0)
+			icon_state = "chamber"
+		if(1)
+			icon_state = "chamber_active"
+		if(2)
+			icon_state = "chamber_active_loaded"
+		if(3)
+			icon_state = "chamber_loaded"
+		if(4)
+			icon_state = "chamber_open"
```
Comment (kriskog) [inline code/modules/research/bepis.dm]: Invert this, set icon_state directly and early return. i.e.
```
/obj/machinery/rnd/bepis/update_icon_state()
	if(panel_open == TRUE)
		icon_state = "chamber_open"
		return
	if (((powered == FALSE) && (banked_cash > 0)) || (banked_cash > 0) && (!is_operational()))
		icon_state = "chamber_loaded"
		return
```
and so on. Then get rid of chamber_status.
Still feels a little messy, you are checking some things again and again, but its a little better.

File: code/modules/research/bepis.dm [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+
+/obj/machinery/rnd/bepis
+	name = "\improper B.E.P.I.S. Chamber"
+	desc = "A high fidelity testing device which unlocks the secrets of the known universe using the two most powerful substances available to man: excessive amounts of electricity and capital."
+	icon = 'icons/obj/machines/bepis.dmi'
+	icon_state = "chamber"
+	layer = ABOVE_MOB_LAYER
+	density = TRUE
+	use_power = IDLE_POWER_USE
+	idle_power_usage = 1500
```
Comment (kriskog) [inline code/modules/research/bepis.dm]: Leave this at 0 (from parent), add active_power_usage = 1500 and swap between use_power = IDLE_POWER_USE and ACTIVE_POWER_USE instead of reassinging this.

File: code/modules/research/bepis.dm [hunk 23]
```diff
@@ -0,0 +1,225 @@
...
+
+//economics defined values, subject to change should anything be too high or low in practice.
+
+
+/obj/machinery/rnd/bepis
+	name = "\improper B.E.P.I.S. Chamber"
+	desc = "A high fidelity testing device which unlocks the secrets of the known universe using the two most powerful substances available to man: excessive amounts of electricity and capital."
+	icon = 'icons/obj/machines/bepis.dmi'
+	icon_state = "chamber"
+	layer = ABOVE_MOB_LAYER
```
Comment (kriskog) [inline code/modules/research/bepis.dm]: This causes it to show above items. Is this for the deer? Dont do this for the deer, and leave it at machines default.

File: code/modules/research/bepis.dm [hunk 24]
```diff
@@ -0,0 +1,225 @@
...
+			var/input = text2num(params["amount"])
+			if(input)
+				banking_amount = input
+		if("toggle_power")
+			if(powered == FALSE)
+				powered = TRUE
+				idle_power_usage = 1500
+			else
+				powered = FALSE
+				idle_power_usage = 0
```
Comment (kriskog) [inline code/modules/research/bepis.dm]:
```
powered = !powered
```
and make this work with active_power_use instead.

File: code/modules/research/bepis.dm [hunk 25]
```diff
@@ -0,0 +1,225 @@
...
+	if(!dropturf) //Check to verify the turf exists and the reward isn't lost somehow.
+		dropturf = drop_location()
+	gauss_major = (gaussian(major_threshold, std) - negative_cash_offset)	//This is the randomized profit value that this experiment has to surpass to unlock a tech.
+	gauss_minor = (gaussian(minor_threshold, std) - negative_cash_offset)	//And this is the threshold to instead get a minor prize.
+	gauss_real = (gaussian(banked_cash, std*inaccuracy_percentage) + positive_cash_offset)	//this is the randomized profit value that your experiment expects to give.
+	say("Real: [gauss_real]. Minor: [gauss_minor]. Major: [gauss_major].")
+	if(gauss_real >= gauss_major) //Major Success.
+		say("Experiment concluded with major success. New technology node discovered on technology disc.")
+		banked_cash = 0
+		new /obj/item/disk/tech_disk/major(dropturf,1)
```
Comment (kriskog) [inline code/modules/research/bepis.dm]: There are no checks to avoid duplicates here. Which means you could be paying for the same tech again and again and again.

File: code/modules/research/bepis.dm [hunk 26]
```diff
@@ -0,0 +1,225 @@
...
+	var/gauss_major = 0
+	var/gauss_minor = 0
+	var/gauss_real = 0
+	if(!dropturf) //Check to verify the turf exists and the reward isn't lost somehow.
+		dropturf = drop_location()
+	gauss_major = (gaussian(major_threshold, std) - negative_cash_offset)	//This is the randomized profit value that this experiment has to surpass to unlock a tech.
+	gauss_minor = (gaussian(minor_threshold, std) - negative_cash_offset)	//And this is the threshold to instead get a minor prize.
+	gauss_real = (gaussian(banked_cash, std*inaccuracy_percentage) + positive_cash_offset)	//this is the randomized profit value that your experiment expects to give.
+	say("Real: [gauss_real]. Minor: [gauss_minor]. Major: [gauss_major].")
+	if(gauss_real >= gauss_major) //Major Success.
```
Comment (kriskog) [inline code/modules/research/bepis.dm]: banked_cash will always be set to 0 after this, do it before the ifs. Same for flick and icon state.

File: code/modules/research/bepis.dm [hunk 27]
```diff
@@ -0,0 +1,225 @@
...
+		say("Experiment concluded with major success. New technology node discovered on technology disc.")
+		banked_cash = 0
+		new /obj/item/disk/tech_disk/major(dropturf,1)
+		flick("chamber_flash",src)
+		update_icon_state()
+		return
+	else if(gauss_real >= gauss_minor) //Minor Success.
```
Comment (kriskog) [inline code/modules/research/bepis.dm]: You are early returning, no need for the else.

File: code/modules/research/bepis.dm [hunk 28]
```diff
@@ -0,0 +1,225 @@
...
+		reward_number = rand(1,2)
+		if(reward_number == 1)
+			new /obj/item/stack/circuit_stack/full(dropturf)
+		if(reward_number == 2)
+			new /obj/item/airlock_painter/decal(dropturf)
+		banked_cash = 0
+		flick("chamber_flash",src)
+		update_icon_state()
+		return
+	else if(gauss_real <= -1)	//Critical Failure
```
Comment (kriskog) [inline code/modules/research/bepis.dm]: else

File: code/modules/research/bepis.dm [hunk 29]
```diff
@@ -0,0 +1,225 @@
...
+		say("ERROR: CRITICAL MACHIME MALFUNCTI- ON. CURRENCY IS NOT CRASH. CANNOT COMPUTE COMMAND: 'make bucks'") //not a typo, for once.
+		banked_cash = 0
+		new /mob/living/simple_animal/deer(dropturf, 1)
+		use_power(500000 * power_saver) //To prevent gambling at low cost and also prevent spamming for infinite deer.
+		flick("chamber_flash",src)
+		update_icon_state()
+		return
+	else	//Minor Failure
```
Comment (kriskog) [inline code/modules/research/bepis.dm]: else, just move all inside this one left

File: code/modules/research/bepis.dm [hunk 30]
```diff
@@ -0,0 +1,225 @@
...
+		say("Attempting to deposit 0 credits. Aborting.")
+		return
+	deposit_value = CLAMP(round(deposit_value, 1), 1, 30000)
+	if(!account)
+		say("Cannot find user account. Aborting.")
+		return
+	if(!account.has_money(deposit_value))
+		say("You do not possess enough credits.")
+		return
+	else
```
Comment (kriskog) [inline code/modules/research/bepis.dm]: early return and else

**Review by kriskog — CHANGES_REQUESTED** [review]
The thing should have a way to remove credits as well.

File: code/game/machinery/doors/airlock_electronics.dm [hunk 31]
```diff
@@ -65,3 +65,55 @@
...
+		if(zero_amount())
+			return
```
[the remark this answers is not in the document]
Comment (kriskog) [reply code/game/machinery/doors/airlock_electronics.dm]: Ignore this one, its probably what he meant after all.

File: code/modules/mob/living/simple_animal/friendly/farm_animals.dm [hunk 32]
```diff
@@ -375,3 +375,34 @@
 		user.visible_message("<span class='notice'>[user] milks [src] using \the [O].</span>", "<span class='notice'>You milk [src] using \the [O].</span>")
 	else
 		to_chat(user, "<span class='warning'>The udder is dry. Wait a bit longer...</span>")
+
+/mob/living/simple_animal/deer
+	name = "doe"
+	desc = "A gentle, peaceful forest animal. How did this get into space?"
+	icon_state = "deer-doe"
```
[the remark this answers is not in the document]
Comment (ArcaneMusic) [reply code/modules/mob/living/simple_animal/friendly/farm_animals.dm]: Can confirm, icondiff bot is super wrong.

File: code/modules/research/bepis.dm (hunk 23 above)
Comment (ArcaneMusic) [reply code/modules/research/bepis.dm]: The problem is that otherwise the player icon sits on-top of the tube itself, since it's a 2 tile tall sprite. I do wish there was a better in-between than this though.
image

Comment (kriskog) [reply code/modules/research/bepis.dm]: Ah, cloner got the same issue doesnt it. Hmm

File: tgui-next/packages/tgui/interfaces/DecalPainter.js [hunk 33]
```diff
@@ -0,0 +1,95 @@
+import { Fragment } from 'inferno';
+import { useBackend } from '../backend';
+import { AnimatedNumber, Box, Button, LabeledList, ProgressBar, Section, Tabs } from '../components';
+
+export const DecalPainter = props => {
+  const { act, data } = useBackend(props);
+  return (
+    <Fragment>
+      <Section title="Decal Style">
```
[the remark this answers is not in the document]
Comment (ArcaneMusic) [reply tgui-next/packages/tgui/interfaces/DecalPainter.js]: Well it sucked, but I did get everything moved over onto a more reactive UI. In the mean time though, the UI's going to be a lot stronger for QOL features at least.

**Comment by ArcaneMusic** [comment]
Alright, changes made, the Main UI now uses swipes for Account registration, you can withdraw credits, a LOT of code clean up, Decal painter's UI has been functionally reworked to the best of my abilities and limitations, so in lou of me saying "I think it's ready" for a third time I'm just going to say it's "probably better now".
image
_New Decal Painter UI_
image
_New BEPIS Machine UI_

File: tgstation.dme [hunk 34]
```diff
@@ -462,7 +462,6 @@
 #include "code\datums\diseases\beesease.dm"
 #include "code\datums\diseases\brainrot.dm"
 #include "code\datums\diseases\cold.dm"
-#include "code\datums\diseases\cold9.dm"
```
Comment (kriskog) [inline tgstation.dme]: Hmm

File: code/game/objects/items/airlock_painter.dm [hunk 35]
```diff
@@ -125,3 +125,114 @@
...
+	decal_list = list(list("Warning Line","warningline"),
+			list("Warning Line Corner","warninglinecorner"),
+			list("Caution Label","caution"),
+			list("Directional Arrows","arrows"),
+			list("Stand Clear Label","stand_clear"),
+			list("Box","box"),
+			list("Box Corner","box_corners"),
+			list("Delivery Marker","delivery"),
+			list("Warning Box","warn_full"))
+	color_list = list("","red","white")
```
Comment (kriskog) [inline code/game/objects/items/airlock_painter.dm]: Just directly set these, no need to do it in Initialize

File: code/modules/research/bepis.dm [hunk 36]
```diff
@@ -0,0 +1,231 @@
...
+	use_power = IDLE_POWER_USE
+	active_power_usage = 1500
+	circuit = /obj/item/circuitboard/machine/bepis
+
+	var/banking_amount = 100
+	var/banked_cash = 0					//stored player cash
+	var/datum/bank_account/account		//payer's account.
+	var/account_name					//name of the payer's account.
+	var/error_cause = null
+	var/powered = FALSE
```
Comment (kriskog) [inline code/modules/research/bepis.dm]: Get rid of this one, check and update use_power instead.

File: code/modules/research/bepis.dm [hunk 37]
```diff
@@ -0,0 +1,231 @@
...
+	//Minor Failure
+	error_cause = pick("attempted to sell grey products to American dominated market.","attempted to sell gray products to British dominated market.","placed wild assumption that PDAs would go out of style.","simulated product #76 damaged brand reputation mortally.","simulated business model resembled 'pyramid scheme' by 98.7%.","product accidently granted override access to all station doors.")
+	say("Experiment concluded with zero product viability. Cause of error: [error_cause]")
+	return
+
+/obj/machinery/rnd/bepis/update_icon_state()
+	if(panel_open == TRUE)
+		icon_state = "chamber_open"
+		return
+	if((powered == TRUE) && is_operational())
```
Comment (kriskog) [inline code/modules/research/bepis.dm]: If you were keeping this, just if(powered) would do the same thing. The opposite would be if(!powered) for false.

File: code/modules/research/bepis.dm [hunk 38]
```diff
@@ -0,0 +1,231 @@
...
+			withdrawcash()
+		if("begin_experiment")
+			if(powered == FALSE)
+				return
+			if(banked_cash == 0)
+				say("Please deposit funds to begin testing.")
+				return
+			calcsuccess()
+			use_power(MACHINE_OPERATION * power_saver) //This thing should eat your APC battery if you're not careful.
+			powered = FALSE	//Shuts off after use in order to get the right visual look. Kinda annoying, but also prevents some possible spam issues.
```
Comment (kriskog) [inline code/modules/research/bepis.dm]: Should happen when you replace powered, but make sure you actually update idle/active here.

**Review by kriskog — CHANGES_REQUESTED** [review]
Some more stuff

File: tgstation.dme (hunk 34 above)
Comment (ArcaneMusic) [reply tgstation.dme]: I was legitimately sure that cold9 was a mistake on my part, whoops
